api/agent/pure_runner.go

```diff
@@ -836,87 +831,78 @@ func (pr *pureRunner) runStatusCall(ctx context.Context) *runner.RunnerStatus {
 	result.Details = string(body)
 	result.Id = c.ID
 
-	if result.Failed {
-		log.Errorf("Status call failure id=%v result=%+v", c.ID, result)
-	} else {
-		log.Debugf("Status call success id=%v result=%+v", c.ID, result)
-	}
-
+	log.Debugf("Finished status call id=%v result=%+v", c.ID, result)
 	return result
 }
 
 // Handles a status call concurrency and caching.
 func (pr *pureRunner) handleStatusCall(ctx context.Context) (*runner.RunnerStatus, error) {
-	var myChan chan struct{}
+	var waitChan chan struct{}
 
-	isWaiter := false
+	isSpawner := false
 	isCached := false
 	now := time.Now()
 
 	pr.status.lock.Lock()
 
-	if now.Before(pr.status.expiry) {
-		// cache is still valid.
-		isCached = true
-	} else if pr.status.wait != nil {
-		// A wait channel is already installed, we must wait
-		isWaiter = true
-		myChan = pr.status.wait
-	} else {
-		// Wait channel is not present, we install a new one.
-		myChan = make(chan struct{})
-		pr.status.wait = myChan
+	isCached = now.Before(pr.status.expiry)
+	if !isCached {
+		if pr.status.wait == nil {
+			pr.status.wait = make(chan struct{})
+			isSpawner = true
+		}
+		waitChan = pr.status.wait
```

Comment:
maybe nitty but it would make the rest of the function a good deal simpler I think if we had a little func that did 895-903 and loaded the cache object and returned it, so that we could do something like:

```go
if isCached {
  return loadCacheObj(pr.status)
}

// now handle the uncached case
```

there's pretty much 3 blocks below here that handle whether it's a spawner or not a cached object, and it can't be a spawner if it's cached either, so I think it will clean up a lot of the conditional blocks and they could just be done at the top level for the !isCached at least, and the spawner case is slightly more straightforward of a condition. it will read more simply... is it cached? yes, okay return or no, do we need to spawn? okay, maybe, then we wait for that result and load it and return it. et voila...

that being said, the code seems correct as i've understood it as is.

Reply:
I've resisted so far to split functions off of this big function to keep/isolate lock code in same block. But that's probably OK, let me see if that version will look simpler.